File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.py
class Solution:
    def maxLength(self, arr: List[str]) -> int:
        def backtrack(ls):
            if len(ls)<=1:
                return ls
            combine = backtrack(ls[1:])
            for i in combine:
                if len(i+ls[0])==len(set(i+ls[0])):
                    combine+=[i+ls[0]]
            combine+=[ls[0]]
            return combine
        
        for i in range(len(arr)-1,-1,-1):
            if len(arr[i])!=len(set(arr[i])):
                arr.pop(i)
        combine = backtrack(arr)
        ans = 0
        for i in combine:
            ans = max(ans, len(i))
            
        return ans

File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    struct permute{
        vector<string> array;
        string cur;
        int idx;
    };
    void dfs(permute tmp){
        for(int i=tmp.idx;i<tmp.array.size();i++){
            string s = tmp.cur+tmp.array[i];
            bool flag=true;
            for(int j=0;j<s.length();j++){
                if(count(s.begin(),s.end(),s[j])>1){
                    flag=false;
                    break;
                }
            }
            if(flag){
                permute next = tmp;
                next.cur+=next.array[i];
                if(next.cur.length()>ans){
                    ans = next.cur.length();
                }
                next.idx = i+1;
                dfs(next);
            }
        }
    }
    int maxLength(vector<string>& arr) {
        permute tmp{arr,"",0};
        dfs(tmp);
        return ans;
    }
private:
    int ans = 0;
};